src/Xamarin.Android.Build.Tasks/Utilities/MarshalMethodsAssemblyRewriter.cs

```diff
@@ -0,0 +1,134 @@
+#if ENABLE_MARSHAL_METHODS
+using System;
+using System.Collections.Generic;
+using System.IO;
+
+using Java.Interop.Tools.Cecil;
+using Mono.Cecil;
+
+namespace Xamarin.Android.Tasks
+{
+	class MarshalMethodsAssemblyRewriter
+	{
+		IDictionary<string, MarshalMethodEntry> methods;
+		ICollection<AssemblyDefinition> uniqueAssemblies;
+		IDictionary <string, HashSet<string>> assemblyPaths;
+
+		public MarshalMethodsAssemblyRewriter (IDictionary<string, MarshalMethodEntry> methods, ICollection<AssemblyDefinition> uniqueAssemblies, IDictionary <string, HashSet<string>> assemblyPaths)
+		{
+			this.methods = methods ?? throw new ArgumentNullException (nameof (methods));
+			this.uniqueAssemblies = uniqueAssemblies ?? throw new ArgumentNullException (nameof (uniqueAssemblies));
+			this.assemblyPaths = assemblyPaths ?? throw new ArgumentNullException (nameof (assemblyPaths));
+		}
+
+		public void Rewrite (DirectoryAssemblyResolver resolver)
+		{
+			var unmanagedCallersOnlyAttributes = new Dictionary<AssemblyDefinition, CustomAttribute> ();
+			foreach (AssemblyDefinition asm in uniqueAssemblies) {
+				unmanagedCallersOnlyAttributes.Add (asm, GetUnmanagedCallersOnlyAttribute (asm, resolver));
+			}
+
+			Console.WriteLine ("Adding the [UnmanagedCallersOnly] attribute to native callback methods and removing unneeded fields+methods");
+			foreach (MarshalMethodEntry method in methods.Values) {
+				Console.WriteLine ($"\t{method.NativeCallback.FullName} (token: 0x{method.NativeCallback.MetadataToken.RID:x})");
+				method.NativeCallback.CustomAttributes.Add (unmanagedCallersOnlyAttributes [method.NativeCallback.Module.Assembly]);
+				method.Connector.DeclaringType.Methods.Remove (method.Connector);
+				method.CallbackField?.DeclaringType.Fields.Remove (method.CallbackField);
+			}
+
+			Console.WriteLine ();
+			Console.WriteLine ("Rewriting assemblies");
+
+			var newAssemblyPaths = new List<string> ();
+			foreach (AssemblyDefinition asm in uniqueAssemblies) {
+				foreach (string path in GetAssemblyPaths (asm)) {
+					var writerParams = new WriterParameters {
+						WriteSymbols = (File.Exists (path + ".mdb") || File.Exists (Path.ChangeExtension (path, ".pdb"))),
+					};
+
+					string output = $"{path}.new";
+					Console.WriteLine ($"\t{asm.Name} => {output}");
+					asm.Write (output, writerParams);
+					newAssemblyPaths.Add (output);
+				}
+			}
+
+			// Replace old versions of the assemblies only after we've finished rewriting without issues, otherwise leave the new
+			// versions around.
+			foreach (string path in newAssemblyPaths) {
+				string target = Path.Combine (Path.GetDirectoryName (path), Path.GetFileNameWithoutExtension (path));
+				MoveFile (path, target);
```

Comment:
We may want/need to rework the build system so that we write files into new/separate directories instead of modifying files in-place, otherwise we'll provoke the incremental build + file sharing gods.

Reply:
I asked @jonathanpeppers about this and he thought it was better to write them in place. We can optimize this in the last PR in the series though